### ISSUE TITLE
Doesn't work... Why ?

### ISSUE THREAD

**Issue by ctoslab** [pdf-utilities/inject_javascript#6]
**Describe the bug**
It doesn't work.

I don't know...

**To Reproduce**
ex. inject-pdf-javascript --pdf "/tmp/empty.pdf" --js "/dir/evil.js"

```bash
root@kali:~/Desktop# inject-pdf-javascript --pdf "/tmp/empty.pdf" --js "/dir/evil.js" --save-path "output/evil.pdf"
Traceback (most recent call last):
  File "/usr/local/bin/inject-pdf-javascript", line 11, in <module>
    load_entry_point('inject-javascript==0.0.7', 'console_scripts', 'inject-pdf-javascript')()
  File "/root/.local/lib/python3.7/site-packages/inject_javascript/cli.py", line 88, in main
    save_path = args.get('save_path'))
  File "/root/.local/lib/python3.7/site-packages/inject_javascript/__init__.py", line 123, in inject_pdf_with_javascript
    pdf_data = self.return_pdf_data(pdf_path = pdf_path)
  File "/root/.local/lib/python3.7/site-packages/inject_javascript/__init__.py", line 53, in return_pdf_data
    with open(pdf_path, 'rb') as fb:
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/empty.pdf'
root@kali:~/Desktop# 

```

**Expected behavior**

**Screenshots**

**Desktop (please complete the following information):**

 - OS: [e.g. Debian]
Kali Linux
 - Version [e.g. 0.0.1]
All versions
**Additional context**

I cannot install it using pip i did:
python3 setup.py install
a lot of problems please fix !

**Comment by S0AndS0** [comment]
Tracing through the error log shows...

```
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/empty.pdf'
```

... which leads to the question. Is there a source PDF at `/tmp/empty.pdf`?

If there is not a source PDF, then it be possible to utilize Chrome web-browser _`Print to PDF`_ to generate a _mostly_ empty PDF by navigating to `about:blank` and using the keyboard short-cut <kbd><kbd>Ctrl</kbd> <kbd>p</kbd></kbd>

**Comment by ctoslab** [comment]
Of course there is i just named it empty.pdf cuz it's blank inside i was just trying if this works...

**Comment by S0AndS0** [comment]
I've done some testing and de-bugging on my end, version `0.0.8` is now out and fixes a few bugs. Updating can be done via...


```Bash
pip3 install --upgrade inject-javascript
```


... However, the only way I could reproduce the errors reported here was by targeting a non-existent file, eg...


```Bash
inject-pdf-javascript --pdf no.pdf --js alert.js --save-path output.pdf
```


```
Traceback (most recent call last):
  File "~/.local/bin/inject-pdf-javascript", line 8, in <module>
    sys.exit(main())
  File "~/.local/lib/python3.6/site-packages/inject_javascript/cli.py", line 88, in main
    save_path = args.get('save_path'))
  File "~/.local/lib/python3.6/site-packages/inject_javascript/__init__.py", line 121, in inject_pdf_with_javascript
    pdf_data = self.return_pdf_data(pdf_path = pdf_path)
  File "~/.local/lib/python3.6/site-packages/inject_javascript/__init__.py", line 53, in return_pdf_data
    with open(pdf_path, 'rb') as fb:
FileNotFoundError: [Errno 2] No such file or directory: 'no.pdf'
```


... I was also able to produce other error output by feeding in a malformed PDF, the `PyPDF2` package did not take kindly to that. So after double-checking that `/tmp/empty.pdf` is a valid file path, eg...

```Bash
ls -ahl "/tmp/empty.pdf"
```


... it would be a good idea to also check that the file content is properly formatted.


______


That all stated, as of version `0.0.8` when targeting a source PDF that does exist no errors are produced, eg...


```Bash
inject-pdf-javascript --pdf blank.pdf --js alert.js --save-path output.pdf 
```


... for completeness here be the plane-text of files used and produced...


**`blank.pdf`**


```PDF
%PDF-1.4
BINARY_BLOB
1 0 obj
<</Creator (Mozilla/5.0 \(X11; Linux x86_64\) AppleWebKit/537.36 \(KHTML, like Gecko\) Chrome/87.0.4280.47 Safari/537.36)
/Producer (Skia/PDF m87)
/CreationDate (D:20201111033656+00'00')
/ModDate (D:20201111033656+00'00')>>
endobj
3 0 obj
<</ca 1
/BM /Normal>>
endobj
4 0 obj
<</Filter /FlateDecode
/Length 108>> stream
BINARY_BLOB
endstream
endobj
2 0 obj
<</Type /Page
/Resources <</ProcSet [/PDF /Text /ImageB /ImageC /ImageI]
/ExtGState <</G3 3 0 R>>>>
/MediaBox [0 0 612 792]
/Contents 4 0 R
/StructParents 0
/Parent 5 0 R>>
endobj
5 0 obj
<</Type /Pages
/Count 1
/Kids [2 0 R]>>
endobj
6 0 obj
<</Type /Catalog
/Pages 5 0 R>>
endobj
xref
0 7
0000000000 65535 f 
0000000015 00000 n 
0000000469 00000 n 
0000000254 00000 n 
0000000291 00000 n 
0000000657 00000 n 
0000000712 00000 n 
trailer
<</Size 7
/Root 6 0 R
/Info 1 0 R>>
startxref
759
%%EOF
```


> Note, `BINARY_BLOB` lines where _sanitized_ of non-printable characters prior posting here. Future readers should utilize Chrome or Firefox print to PDF as described previously to reproduce a _mostly_ empty PDF.


---


**`alert.js`**


```JavaScript
var hello = function (arg) {
    window.alert("Hello " + arg + "!");
};
hello('world');

```


---


**`output.pdf`**

```PDF
%PDF-1.3
1 0 obj
<<
/Type /Pages
/Count 1
/Kids [ 3 0 R ]
>>
endobj
2 0 obj
<<
/Producer (PyPDF2)
>>
endobj
3 0 obj
<<
/Type /Page
/Resources <<
/ProcSet [ /PDF /Text /ImageB /ImageC /ImageI ]
/ExtGState <<
/G3 null
>>
>>
/MediaBox [ 0 0 612 792 ]
/Contents null
/StructParents 0
/Parent 1 0 R
>>
endobj
4 0 obj
<<
/Type /Action
/S /JavaScript
/JS (var hello = function (arg) {
    window.alert("Hello " + arg + "!");
};
hello('world');

)
>>
endobj
5 0 obj
<<
/JavaScript <<
/Names [ (49d7149d\055bbdf\0554c46\0558da5\0558ff44205940d) 4 0 R ]
>>
>>
endobj
6 0 obj
<<
/Type /Catalog
/Pages 1 0 R
/OpenAction 4 0 R
/Names <<
/JavaScript <<
/Names [ (49d7149d\055bbdf\0554c46\0558da5\0558ff44205940d) 4 0 R ]
>>
>>
>>
endobj
xref
0 7
0000000000 65535 f 
0000000009 00000 n 
0000000068 00000 n 
0000000108 00000 n 
0000000304 00000 n 
0000000450 00000 n 
0000000557 00000 n 
trailer
<<
/Size 7
/Root 6 0 R
/Info 2 0 R
>>
startxref
723
%%EOF
```

**Comment by ctoslab** [comment]
Thanks now it's working, but what if i want to run calc.exe on launch cuz i did a try to inject the js code into pdf but when i open it nothing happen.

This is the exemple of the JS code that i injected 
```
var WshShell = new ActiveXObject("WScript.Shell");
 WshShell.Run("calc.exe");
```

**Comment by S0AndS0** [comment]
You're very welcome! And I'm glad that the project is now functioning as intended.

As for running JavaScript within a PDF, it depends upon the reader. Currently the Adobe PDF reader is the most feature-full, here's a link the their [API docs](https://www.adobe.com/content/dam/acom/en/devnet/acrobat/pdfs/js_api_reference.pdf). Chrome web-browser has some but not all of the same functionality. And the Firefox web-browser has almost no JavaScript capabilities; last I tested the only thing I got working on Firefox was opening a _"Print Preview"_ dialog.

Executing system commands from JavaScript within a PDF is not something that I've experimented much with, however, a search of [StackOverflow](https://stackoverflow.com/questions/7027698/how-to-open-a-pdf-with-wscript-shell) shows that it may be worth prepending `"start"` to the command, eg...

```JavaScript
var wshShell = new ActiveXObject("WScript.Shell");
wshShell.Run("start calc.ext", 1, false);
```


---

Bonus link I ran across, [The Acrobat JavaScript Console](https://acrobatusers.com/tutorials/javascript_console/), that shows how to enable a debugging console within Adobe PDF reader, which may allow for a swifter development cycle.


[MDN ActiveXObject](https://developer.mozilla.org/en-US/docs/Archive/Web/JavaScript/Microsoft_Extensions/ActiveXObject)
